File: core/src/esrunner/ESRunner.java
package esrunner;

import com.github.emulio.yaml.YamlReaderHelper;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Properties;

public class ESRunner {

	public ESRunner(List<String> args) {
		System.out.println("Initializing ESRunner.");

		printProperties("System", System.getProperties());

		System.out.println("   _____ ____  ____                                \n"
				+ "  | ____/ ___||  _ \\ _   _ _ __  _ __   ___ _ __   \n"
				+ "  |  _| \\___ \\| |_) | | | | '_ \\| '_ \\ / _ \\ '__|  \n"
				+ "  | |___ ___) |  _ <| |_| | | | | | | |  __/ |     \n"
				+ "  |_____|____/|_| \\_\\\\__,_|_| |_|_| |_|\\___|_|     \n");

	}

	private void printProperties(final String name, final Properties properties) {
		System.out.println("===========================================");
		System.out.println(name + " Properties: ");
		System.out.println("===========================================");
		for (Entry<Object, Object> entry : properties.entrySet()) {
			System.out.printf("%s: %s\n", entry.getKey(), entry.getValue());
		}
		System.out.println("===========================================");
	}

	public static void main(String[] args) {
		new ESRunner(Arrays.asList(args)).run();
	}

	private void run() {

		try {
			final Map<Object, Object> esConfig = readConfig();
			overrideEmulationStationFile(esConfig);

		} catch (Throwable t) {
			if (t.getMessage() != null) {
				System.out.println("\n\n[ERROR] " + t.getMessage());
			}

			System.out.println("\n\n\n==========================================================");
			t.printStackTrace(System.out);
		}
	}

	private Map<Object, Object> readConfig() {
		final File esRunnerYaml = new File("esrunner.yaml");
		
		return YamlReaderHelper.parse(esRunnerYaml);
	}

	

	private void overrideEmulationStationFile(final Map<Object, Object> esConfig) {
		System.out.println("Overriding emulationstation config files");
		
		validateConfig(esConfig);
		
		final String es_systemsContent = generateESSystems(esConfig);
		
		final String esHome = (String) esConfig.get("emulationstation.home");
		final String esRunnerHome = (String) esConfig.get("esrunner.home");
		
		final File es_systemsFile = new File(esHome, "es_systems.cfg");
		
		writeFile(es_systemsContent, es_systemsFile);
		
		final Object esCommandObject = esConfig.get("emulationstation.command");
		final ProcessBuilder esProcessBuilder;
		if (esCommandObject instanceof List) {
			@SuppressWarnings("unchecked")
			final List<String> command = (List<String>) esCommandObject;
			esProcessBuilder = new ProcessBuilder(command.toArray(new String[command.size()]));
		} else {
			esProcessBuilder = new ProcessBuilder((String) esCommandObject);
		}
		
		System.out.println("Preparing to start emulationstation");
		
		final Map<String, String> environment = esProcessBuilder.environment();
		environment.put("HOME", esRunnerHome);
		
		System.out.println("Executing emulationstation.. ");
		try {
			final Process process = esProcessBuilder.start();
			
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
		
	}
	
	private static void writeFile(final String content, final File file) {
		System.out.printf("Writing file: [%s] \n", file.getAbsolutePath());
		try (final FileOutputStream fos = new FileOutputStream(file)) {
			fos.write(content.getBytes());
			fos.flush();
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
		System.out.println("Writing done");
	}

	private String generateESSystems(final Map<Object, Object> esConfig) {
		final StringBuilder sb = new StringBuilder();
		sb.append("<systemList>\n");
		
		@SuppressWarnings("unchecked")
		final Map<String, Object> systems = (Map<String, Object>) esConfig.get("systems");
		for (Entry<String, Object> entry : systems.entrySet()) {
		
			final String systemName = entry.getKey();
			System.out.println("Generating config for: " + systemName);
			 
			@SuppressWarnings("unchecked")
			final Map<String, Object> systemMap = (Map<String, Object>) entry.getValue();
			
			sb.append("\t<system>\n");
			sb.append(String.format("\t\t<name>%s</name>\n", systemName));
			sb.append(String.format("\t\t<platform>%s</platform>\n", systemName));
			sb.append(String.format("\t\t<theme>%s</theme>\n", systemName));
			
			final String platformName = (String) systemMap.get("platform.name");
			sb.append(String.format("\t\t<fullname>%s</fullname>\n", platformName));
			
			final Object romsExtensionsObject = systemMap.get("roms.extensions");
			sb.append(String.format("\t\t<extension>"));
			if (romsExtensionsObject instanceof List) {
				@SuppressWarnings("unchecked")
				final List<String> romsExtensions = (List<String>) romsExtensionsObject;
					
				for (String extension : romsExtensions) {
					sb.append(extension.toLowerCase()).append(" ").append(extension.toUpperCase()).append(" ");
				}
				sb.setLength(sb.length() -1);
			} else {
				final String extension = (String) romsExtensionsObject;
				sb.append(extension.toLowerCase()).append(" ").append(extension.toUpperCase());
			}
			sb.append("</extension>\n");
			
			final String romsPath = (String) systemMap.get("roms.path");
			sb.append(String.format("\t\t<path>%s</path>\n", romsPath));
			
			final Object runCommandObject = systemMap.get("run.command");
			sb.append(String.format("\t\t<command>"));
			if (runCommandObject instanceof List) {
				@SuppressWarnings("unchecked")
				final List<String> runCommands = (List<String>) runCommandObject;
					
				for (String command: runCommands) {
					if (command.indexOf(' ') != -1) {
						sb.append("\"").append(command).append("\" ");
					} else {
						sb.append(command).append(" ");
					}
				}
				sb.setLength(sb.length() -1);
			} else {
				sb.append((String) runCommandObject);
			}
			sb.append("</command>\n");
			sb.append("\t</system>\n");
		}		
		sb.append("</systemList>");
		
		
		final String es_systemsContent = sb.toString();
		
		return es_systemsContent;
	}

	private void validateConfig(final Map<Object, Object> esConfig) {
		checkValue("You must provide a systems on your yaml file.", esConfig.get("systems"));
		checkValue("You must provide a emulationstation.home on your yaml file.", esConfig.get("emulationstation.home"));
		checkValue("You must provide a emulationstation.command on your yaml file.", esConfig.get("emulationstation.command"));
		checkValue("You need to specify at least a system.", esConfig.get("systems"));
	}

	

	private void checkValue(String message, Object value) {
		if (value == null) {
			throw new RuntimeException(message);
		}
	}

	

}

File: core/src/main/esrunner/ESRunner.kt
package esrunner

import com.github.emulio.yaml.YamlReaderHelper
import mu.KotlinLogging
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * This is an util class used in the past on a smaller project
 * that manages all the bad configs in the xml files from EmulationStation
 *
 * This was used in another script/project to make EmulationStation usage
 * much more easy to configure and now part of this implementation is used
 * in the Emulio project
 *
 * A lot of code here can be changed and upgraded and we can break compatibility
 * in complete with EmulationStation for now.
 */
class ESRunner : Runnable {

    val logger = KotlinLogging.logger { }

    init {
        logger.info { "Initializing ESRunner." }
        printProperties()
        printMotd()
    }

    private fun printMotd() {
        logger.info {
            """
             _______  _______  ______    __   __  __    _  __    _  _______  ______   
            |       ||       ||    _ |  |  | |  ||  |  | ||  |  | ||       ||    _ |  
            |    ___||  _____||   | ||  |  | |  ||   |_| ||   |_| ||    ___||   | ||  
            |   |___ | |_____ |   |_||_ |  |_|  ||       ||       ||   |___ |   |_||_ 
            |    ___||_____  ||    __  ||       ||  _    ||  _    ||    ___||    __  |
            |   |___  _____| ||   |  | ||       || | |   || | |   ||   |___ |   |  | |
            |_______||_______||___|  |_||_______||_|  |__||_|  |__||_______||___|  |_|
            
            """.trimIndent()
        }
    }

    private fun printProperties() {
        logger.info { "===========================================" }
        logger.info { "System Properties: " }
        logger.info { "===========================================" }
        System.getProperties().forEach { key, value ->
            logger.info { "$key: $value" }
        }
        logger.info { "===========================================" }
    }

    override fun run() {
        try {
            overrideEmulationStationFile(readConfig())
        } catch (t: Throwable) {
            handleThrowable(t)
        }
    }

    private fun handleThrowable(t: Throwable) {
        logger.error(t) { "Internal error ocurred" }
    }

    private fun readConfig(): Map<Any, Any> {
        val esRunnerYaml = File("esrunner.yaml")
        return YamlReaderHelper.parse(esRunnerYaml)
    }

    private fun overrideEmulationStationFile(esConfig: Map<Any, Any>) {
        logger.info { "Overriding EmulationStation config files" }

        validateConfig(esConfig)

        @Suppress("LocalVariableName")
        val es_systemsContent = generateESSystems(esConfig)

        val esHome = esConfig["EmulationStation.home"] as String
        val esRunnerHome = esConfig["esrunner.home"] as String

        @Suppress("LocalVariableName")
        val es_systemsFile = File(esHome, "es_systems.cfg")

        writeFile(es_systemsContent, es_systemsFile)

        val esCommandObject = esConfig["EmulationStation.command"]
        val esProcessBuilder: ProcessBuilder

        esProcessBuilder = if (esCommandObject is List<*>) {
            getListProcessBuilder(esCommandObject)
        } else {
            getProcessBuilder(esCommandObject)
        }

        logger.info { "Preparing to start EmulationStation" }

        val environment = esProcessBuilder.environment()
        environment["HOME"] = esRunnerHome

        logger.info { "Executing EmulationStation.. " }
        try {
            esProcessBuilder.start()
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    }

    private fun getProcessBuilder(esCommandObject: Any?) = ProcessBuilder(esCommandObject as String)

    private fun getListProcessBuilder(esCommandObject: Any?): ProcessBuilder {
        @Suppress("UNCHECKED_CAST")
        val command = esCommandObject as List<String>
        return ProcessBuilder(*command.toTypedArray())
    }

    private fun generateESSystems(esConfig: Map<Any, Any>): String {
        return StringBuilder()
            .append("<systemList>\n")
            .appendSystems(esConfig)
            .append("</systemList>").toString()
    }

    private fun StringBuilder.appendSystems(esConfig: Map<Any, Any>): StringBuilder {
        getSystems(esConfig).forEach { (systemName, value) ->
            this.appendSystem(systemName, value)
        }
        return this
    }

    @Suppress("UNCHECKED_CAST")
    private fun getSystems(esConfig: Map<Any, Any>) =
            esConfig["systems"] as Map<String, Any>

    private fun StringBuilder.appendSystem(systemName: String, value: Any) {
        logger.info { "Generating config for: $systemName" }

        val systemMap = getSystemMap(value)

        this.append("\t<system>\n")
                .appendName(systemName)
                .appendPlatform(systemName)
                .appendTheme(systemName)
                .appendPlatformName(systemMap)
                .appendExtension(systemMap)
                .appendPath(systemMap)
                .appendCommand(systemMap)
            .append("\t</system>\n")
    }

    private fun StringBuilder.appendCommand(systemMap: Map<String, Any>): StringBuilder {
        val runCommandObject = systemMap["run.command"]
        this.append(String.format("\t\t<command>"))
        if (runCommandObject is List<*>) {
            @Suppress("UNCHECKED_CAST")
            val runCommands = runCommandObject as List<String>

            for (command in runCommands) {
                if (command.indexOf(' ') != -1) {
                    this.append("\"").append(command).append("\" ")
                } else {
                    this.append(command).append(" ")
                }
            }
            this.setLength(this.length - 1)
        } else {
            this.append(runCommandObject as String)
        }
        this.append("</command>\n")

        return this
    }

    private fun StringBuilder.appendPath(systemMap: Map<String, Any>): StringBuilder {
        return this.append(String.format("\t\t<path>%s</path>\n", systemMap["roms.path"] as String))
    }

    private fun StringBuilder.appendExtension(systemMap: Map<String, Any>): StringBuilder {
        val romsExtensionsObject = systemMap["roms.extensions"]
        this.append(String.format("\t\t<extension>"))
        if (romsExtensionsObject is List<*>) {
            this.appendExtensionsList(romsExtensionsObject)
        } else {
            this.appendExtension(romsExtensionsObject)
        }
        this.append("</extension>\n")

        return this
    }

    private fun StringBuilder.appendExtension(romsExtensionsObject: Any?) {
        val extension = romsExtensionsObject as String
        this.append(extension.toLowerCase()).append(" ").append(extension.toUpperCase())
    }

    private fun StringBuilder.appendExtensionsList(romsExtensionsObject: Any?) {
        @Suppress("UNCHECKED_CAST")
        val romsExtensions = romsExtensionsObject as List<String>
        romsExtensions.forEach { extension ->
            this.append(extension.toLowerCase()).append(" ").append(extension.toUpperCase()).append(" ")
        }
        this.setLength(this.length - 1)
    }

    @Suppress("UNCHECKED_CAST")
    private fun getSystemMap(value: Any): Map<String, Any> {
        return value as Map<String, Any>
    }

    private fun StringBuilder.appendPlatformName(systemMap: Map<String, Any>): StringBuilder {
        return this.append(String.format("\t\t<fullname>%s</fullname>\n", systemMap["platform.name"] as String))
    }

    private fun StringBuilder.appendTheme(systemName: String): StringBuilder {
        return this.append(String.format("\t\t<theme>%s</theme>\n", systemName))
    }

    private fun StringBuilder.appendPlatform(systemName: String): StringBuilder {
        return this.append(String.format("\t\t<platform>%s</platform>\n", systemName))
    }

    private fun StringBuilder.appendName(systemName: String): StringBuilder {
        return this.append(String.format("\t\t<name>%s</name>\n", systemName))
    }

    private fun validateConfig(esConfig: Map<Any, Any>) {
        checkValue("You must provide a systems on your yaml file.", esConfig["systems"])
        checkValue("You must provide a EmulationStation.home on your yaml file.", esConfig["EmulationStation.home"])
        checkValue("You must provide a EmulationStation.command on your yaml file.", esConfig["EmulationStation.command"])
        checkValue("You need to specify at least a system.", esConfig["systems"])
    }

    private fun checkValue(message: String, value: Any?) {
        if (value == null) {
            throw RuntimeException(message)
        }
    }

    private fun writeFile(content: String, file: File) {
        System.out.printf("Writing file: [%s] \n", file.absolutePath)
        try {
            FileOutputStream(file).use { fos ->
                fos.write(content.toByteArray())
                fos.flush()
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        logger.info { "Writing done" }
    }

}

fun main() {
    ESRunner().run { }
}